src/hooks/connectWeb3/useActiveWeb3React.tsx

```diff
@@ -0,0 +1,120 @@
+
+export default function useActiveWeb3React() {
+  const { web3 } = useContext(Web3Context)
+  return web3
+}
+
+export function useUpdateActiveWeb3ReactCallback() {
+  const { updateWeb3 } = useContext(Web3Context)
+  return updateWeb3
+}
+
+export function getNetwork(jsonRpcEndpoint?: string | JsonRpcProvider) {
+  if (jsonRpcEndpoint) {
+    let connector, hooks
+    if (JsonRpcProvider.isProvider(jsonRpcEndpoint)) {
+      ;[connector, hooks] = initializeConnector((actions) => new JsonRpcConnector(actions, jsonRpcEndpoint))
+    } else {
+      ;[connector, hooks] = initializeConnector((actions) => new Url(actions, jsonRpcEndpoint))
+    }
+    connector.activate()
+    return { connector, hooks }
+  }
+  return EMPTY_STATE
+}
+
+function getWallet(provider?: JsonRpcProvider | Eip1193Provider) {
+  if (provider) {
+    let connector, hooks
+    if (JsonRpcProvider.isProvider(provider)) {
+      ;[connector, hooks] = initializeConnector((actions) => new JsonRpcConnector(actions, provider))
+    } else if (JsonRpcProvider.isProvider((provider as any).provider)) {
+      throw new Error('Eip1193Bridge is experimental: pass your ethers Provider directly')
+    } else {
+      ;[connector, hooks] = initializeConnector((actions) => new EIP1193(actions, provider))
+    }
+    connector.activate()
+    return { connector, hooks }
+  }
+  return EMPTY_STATE
+}
+
+interface ActiveWeb3ProviderProps {
+  jsonRpcEndpoint?: string | JsonRpcProvider
+  provider?: Eip1193Provider | JsonRpcProvider
+}
+
+export function ActiveWeb3Provider({
+  jsonRpcEndpoint,
+  provider,
+  children,
+}: PropsWithChildren<ActiveWeb3ProviderProps>) {
+  const network = useMemo(() => getNetwork(jsonRpcEndpoint), [jsonRpcEndpoint])
+  const wallet = useMemo(() => getWallet(provider), [provider])
+
+  // eslint-disable-next-line prefer-const
+  let { connector, hooks } = wallet.hooks.useIsActive() || network === EMPTY_STATE ? wallet : network
+  let accounts = hooks.useAccounts()
+  let account = hooks.useAccount()
+  let activating = hooks.useIsActivating()
+  let active = hooks.useIsActive()
+  let chainId = hooks.useChainId()
+  let error = hooks.useError()
+  let library = hooks.useProvider()
+
+  const web3 = useMemo(() => {
+    if (connector === EMPTY || !(active || activating)) {
+      return EMPTY_WEB3
+    }
+    return { connector, library, chainId, accounts, account, active, activating, error }
+  }, [account, accounts, activating, active, chainId, connector, error, library])
+
+  const updateWeb3 = (updateContext: Web3ContextType) => {
+    connector = updateContext.connector
+    accounts = updateContext.accounts
+    account = updateContext.account
+    activating = updateContext.activating ?? false
+    active = updateContext.active ?? false
+    chainId = updateContext.chainId
+    error = updateContext.error
+    library = updateContext.library as Web3Provider
+  }
+
+  const value = { web3, updateWeb3 }
```

Comment:
I think it'd be reasonable to just skip the assignment here and drop the object into the provider prop as you create it:

`<Web3Context.Provider value={{ web3, updateWeb3 }}>{children}</Web3Context.Provider>`